### PR TITLE
feat: webui create user prompts for password [DET-10221]

### PR DESCRIPTION
**Pull request by jesse-amano-hpe**
## Ticket
[DET-10221]  feat: prompt for password when creating a user in web UI

## Description
BREAKING CHANGE: adds a password field to the "add user" and "edit user" modals. The password is required when creating a new user; if left blank while editing an existing user, their password will remain unchanged.

image

image

## Test Plan
The unit test for this modal has been updated to ensure a password is used and the form submits properly.

To test manually, log in as an admin, and ensure that an invalid password cannot be submitted when adding a new user, and a valid one can. With an SSO provider enabled, marking a created user as Remote should remove the password prompt and requirements.

Known issues:
- The "User Password required" error message may display redundantly and repeatedly, especially when deleting and retyping a password multiple times. This seems to be a bug in the React component for required fields. It also happens on the existing "change my own password" workflow in the User Settings drawer.

## Checklist

- [x] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-10221]: https://hpe-aiatscale.atlassian.net/browse/DET-10221?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by keita-determined — COMMENTED** [review]
just noticed warnings. (im a non-reviewer :))


https://github.com/determined-ai/determined/assets/109113616/b0c28851-d2c4-4bd4-bc7f-c31949c88387

File: webui/react/src/e2e/fixtures/user.fixture.ts [hunk 1]
```diff
@@ -51,16 +55,17 @@ export class UserFixture {
     username = safeName('test-user'),
     displayName,
     isAdmin,
+    password = 'testPassword1',
```
Comment (djanicekpach) [inline webui/react/src/e2e/fixtures/user.fixture.ts:58]: Can we pull this from an env variable in circle CI? The playwright context should probably be the context for it. We should avoid hardcoding passwords in these tests so that they are environment agnostic.

File: webui/react/src/e2e/fixtures/user.fixture.ts [hunk 2]
```diff
@@ -24,13 +25,16 @@ export class UserFixture {
...
 
-  async fillUserForm({ username, displayName, isAdmin }: UserArgs): Promise<void> {
+  async fillUserForm({ username, displayName, isAdmin, password }: UserArgs): Promise<void> {
     if (username !== undefined) {
       await this.userManagementPage.createUserModal.username.pwLocator.fill(username);
     }
     if (displayName !== undefined) {
       await this.userManagementPage.createUserModal.displayName.pwLocator.fill(displayName);
     }
+    if (password !== undefined) {
```
Comment (djanicekpach) [inline webui/react/src/e2e/fixtures/user.fixture.ts:35]: I believe tests are failing because the UI doesn't allow blank passwords on edit. Is that OK? Do users need to update their password whenever they edit? That seems like it should only be required on creation to me.


If we are sure about this behavior then the edit test needs an update.
Line ref: https://github.com/determined-ai/determined/blob/985daeb7056cc9343ab8614d5fb4dc31b549855e/webui/react/src/e2e/tests/userManagement.spec.ts#L114

**Review by djanicekpach — APPROVED** [review]

**Review by djanicekpach — CHANGES_REQUESTED** [review]
I see why the e2 tests fail, but I think the app behavior is a bit odd there. Let me know your thoughts.

File: webui/react/src/e2e/fixtures/user.fixture.ts [hunk 3]
```diff
@@ -14,6 +14,7 @@ interface UserArgs {
   username?: string;
   displayName?: string;
   isAdmin?: boolean;
+  password?: string;
```
Comment (JComins000) [inline webui/react/src/e2e/fixtures/user.fixture.ts:17]: we want password in our `User` interface as well, right?

File: webui/react/src/e2e/fixtures/user.fixture.ts (hunk 1 above)
Comment (JComins000) [reply webui/react/src/e2e/fixtures/user.fixture.ts:58]: I don't like seeing a default here. it's not a required field, is it?

Comment (JComins000) [reply webui/react/src/e2e/fixtures/user.fixture.ts:58]: if it's required upon create, I'd suggest a new `EditArgs` interface

File: webui/react/src/e2e/models/components/CreateUserModal.ts [hunk 4]
```diff
@@ -25,6 +25,10 @@ export class CreateUserModal extends Modal {
     parent: this.body,
     selector: '[data-testid="isRemote"] button',
   });
+  readonly password: BaseComponent = new BaseComponent({
```
Comment (JComins000) [inline webui/react/src/e2e/models/components/CreateUserModal.ts:28]: can you include the form area so we can get PASSWORD_RULES and check for those appearing properly? it could be in another PR if you want

File: webui/react/src/e2e/fixtures/user.fixture.ts (hunk 3 above)
Comment (JComins000) [reply webui/react/src/e2e/fixtures/user.fixture.ts:17]: maybe `EditArgs` can have optional password, and
```
interface UserArgs extends EditArgs
```
with password required?

File: webui/react/src/e2e/fixtures/user.fixture.ts (hunk 2 above)
Comment (jesse-amano-hpe) [reply webui/react/src/e2e/fixtures/user.fixture.ts:35]: No, I agree, the intent was that a blank password on the edit modal should mean "keep the password the same" and it shouldn't be required if the user exists already.

File: webui/react/src/e2e/fixtures/user.fixture.ts (hunk 3 above)
Comment (jesse-amano-hpe) [reply webui/react/src/e2e/fixtures/user.fixture.ts:17]: I don't think we want this in the `User` interface; I may be wrong, but based on how it's used it seems like it represents a server response, and the server wouldn't put the user password in a response.

Comment (JComins000) [reply webui/react/src/e2e/fixtures/user.fixture.ts:17]: i'm just using it as an "internal test representation" of a user. think of it as "everything an admin would think of" when they think of a user. for example, if i'm gonna be logging in with that user later, i'm gonna need to know it's password.

**Comment by jesse-amano-hpe** [comment]
All right, thanks to @JComins000 e2e tests are passing, and with his tests merged in there's pretty good coverage for what happens to a user after their password is set or changed with this new workflow as well. @djanicekpach would you mind taking another look when you have time?

**Review by JComins000 — APPROVED** [review]
good stuff! help me out with PRs to readme.md and password validation later :-)

**Review by gt2345 — COMMENTED** [review]
Password prompts work as expected 👍  Should we mention this in release notes? 

Also want to mention that usually there are two inputs of passwords, something like
Screenshot 2024-04-29 at 3 17 54 PM 
I don't feel too strongly about this, it's up to you and/or design.

One thing I noticed is when rabc enabled, the password field does not show up when remote is on and off, is this expected? Shouldn't we allow setting password when remote is off? 
Screenshot 2024-04-29 at 3 09 33 PM

**Comment by jesse-amano-hpe** [comment]
@gt2345 Thanks for the review!

> Password prompts work as expected 👍 Should we mention this in release notes?

Added.

> Also want to mention that usually there are two inputs of passwords[.] I don't feel too strongly about this, it's up to you and/or design.

I'd personally prefer to go without, since there's a visibility toggle on the field already; not sure who else I'd confirm design with here, but I'll double-check with my team.

> One thing I noticed is when rabc enabled, the password field does not show up when remote is on and off, is this expected? Shouldn't we allow setting password when remote is off?

Great catch, I missed parentheses over a group of conditions, but on another go through it, the rbac check was redundant anyway. Should be fixed now. Thanks!

**Review by gt2345 — COMMENTED** [review]
Thanks for the fix. 

I noticed that with rbac enabled, when editing modal opens for remote user, the password field would show up even though it shouldn't 
Screenshot 2024-04-30 at 12 48 20 PM

It also seems like we can submit password change for remote user, and backend would throw an error. 

- Using OSS, we can submit password changes for remote user. I'm not sure this would happen during production though. 
- Using rbac, we can first change a remote user to non-remote user, then submitting password change would trigger the error. I think this maybe related to the backend.

**Comment by jesse-amano-hpe** [comment]
> Thanks for the fix.
> 
> It also seems like we can submit password change for remote user, and backend would throw an error.
> 
> * Using OSS, we can submit password changes for remote user. I'm not sure this would happen during production though.
> * Using rbac, we can first change a remote user to non-remote user, then submitting password change would trigger the error. I think this maybe related to the backend.

@gt2345 looks like you caught another typo. https://github.com/determined-ai/determined/pull/9240/commits/df77cfcfc8c2053dae63057a69000877f3c7d1ba fixes it.

**Review by gt2345 — APPROVED** [review]
Password prompts work as expected 👍
